### PR TITLE
proactiveConnectionConcurrency

### PR DESCRIPTION
**Pull request by xinlian12**
Test

### CODE REVIEW

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java [hunk 1]
```diff
@@ -116,23 +121,56 @@ public void close() throws IOException {
         }
     }
 
-    public void recordOpenConnectionsAndInitCachesCompleted() {
-        this.isConcurrencySwitchFlow.set(false);
-        this.aggressivenessHint.set(ConnectionOpenFlowAggressivenessHint.DEFENSIVE);
-        this.reInstantiateOpenConnectionsPublisherAndSubscribe(true);
+    public void recordOpenConnectionsAndInitCachesCompleted(List<CosmosContainerIdentity> containerIdentities) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java:124]: probably should be synchornized

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java [hunk 2]
```diff
@@ -116,23 +127,73 @@ public void close() throws IOException {
...
+
     }
 
-    public void recordOpenConnectionsAndInitCachesStarted() {
-        this.isConcurrencySwitchFlow.set(true);
-        this.aggressivenessHint.set(ConnectionOpenFlowAggressivenessHint.AGGRESSIVE);
-        this.reInstantiateOpenConnectionsPublisherAndSubscribe(false);
+    public void recordOpenConnectionsAndInitCachesStarted(List<CosmosContainerIdentity> cosmosContainerIdentities) {
+
+        endpointsUnderMonitorMapWriteLock.lock();
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java:160]: Why ReadWriteLock?

All you need is synchronized (this) {
xxx
}

 if (shouldReInstantiatePublisher) {
            this.aggressivenessHint.set(ConnectionOpenFlowAggressivenessHint.AGGRESSIVE);
            this.reInstantiateOpenConnectionsPublisherAndSubscribe(false);
        }

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - just one question/comment around the lock mechnism chosen

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java (hunk 2 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java]: updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java [hunk 3]
```diff
@@ -710,6 +710,7 @@ public <T> CosmosPagedFlux<T> queryItems(String query, Class<T> classType) {
      *
      *  @return Mono of Void.
      */
+    @Deprecated
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:713]: Please add link to the new PAI to-be-used instead

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - except for link to new APIs to-be-used on deprecated container initialization methods

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks!

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java [hunk 4]
```diff
@@ -940,14 +942,21 @@ public CosmosAsyncClient buildAsyncClient() {
...
+            cosmosAsyncClient.recordOpenConnectionsAndInitCachesStart(proactiveContainerInitConfig.getCosmosContainerIdentities());
+
             Duration aggressiveProactiveConnectionEstablishmentTimeWindow = proactiveContainerInitConfig
                     .getAggressiveProactiveConnectionEstablishmentDuration();
             if (aggressiveProactiveConnectionEstablishmentTimeWindow != null) {
                 cosmosAsyncClient.openConnectionsAndInitCaches(aggressiveProactiveConnectionEstablishmentTimeWindow);
             } else {
                 cosmosAsyncClient.openConnectionsAndInitCaches();
             }
+
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosClientBuilder.java:955]: I like how you decoupled `ProactiveOpenConnectionsProcessor` from `warmUpDuration`.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java [hunk 5]
```diff
@@ -25,78 +27,46 @@
...
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
 
 public final class ProactiveOpenConnectionsProcessor implements Closeable {
 
     private static final Logger logger = LoggerFactory.getLogger(ProactiveOpenConnectionsProcessor.class);
     private Sinks.Many<OpenConnectionTask> openConnectionsTaskSink;
     private Sinks.Many<OpenConnectionTask> openConnectionsTaskSinkBackUp;
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java:39]: Nit: `openConnectionsTaskSinkBackUp` can be removed.

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java]: updated

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java (hunk 3 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java:714]: added

**Comment by jeet1995** [comment]
Looks good to me.

**Pull request merged**